### ISSUE TITLE
Mongooseim 1.6.1 Crash on HTTP Rest Auth

### ISSUE THREAD

**Issue by jonathanve** [esl/MongooseIM#626]
Hello, there

Mongooseim 1.6.1 is crashing with the next log:

```erlang
2015-12-22 12:54:18 =ERROR REPORT====
** State machine <0.1596.0> terminating 
** Last event in was {xmlstreamelement,{xmlel,<<"auth">>,[{<<"xmlns">>,<<"urn:ietf:params:xml:ns:xmpp-sasl">>},{<<"mechanism">>,<<"PLAIN">>}],[{xmlcdata,<<"AUTH_DATA">>}]}}
** When State == wait_for_feature_request
**      Data  == {state,{socket_state,gen_tcp,#Port<0.5963>,<0.1595.0>},ejabberd_socket,#Ref<0.0.0.63723>,false,<<"583280513">>,{sasl_state,<<"jabber">>,<<"x.twnel.net">>,<<>>,#Fun<ejabberd_c2s.2.34911987>,#Fun<ejabberd_c2s.3.34911987>,#Fun<ejabberd_c2s.4.34911987>,undefined,undefined},c2s,c2s_shaper,{false,0},false,false,false,[verify_none],false,undefined,<<>>,<<"x.twnel.net">>,<<>>,undefined,{pres_t,0},{pres_f,0},{pres_a,0},{pres_i,0},[],undefined,undefined,undefined,false,{userlist,none,[],false},unknown,undefined,{{10,3,1,111},50590},[],<<>>,false,0,undefined,0,[],0,100,1,600,undefined,undefined}
** Reason for termination = 
** {noproc,{gen_server,call,[<0.1595.0>,reset_stream]}}
2015-12-22 12:54:18 =CRASH REPORT====
  crasher:
    initial call: gen:init_it/6
    pid: <0.1596.0>
    registered_name: []
    exception exit: {{noproc,{gen_server,call,[<0.1595.0>,reset_stream]}},[{p1_fsm_old,terminate,7,[{file,"src/p1_fsm_old.erl"},{line,695}]},{proc_lib,init_p_do_apply,3,[{file,"proc_lib.erl"},{line,237}]}]}
    ancestors: [ejabberd_c2s_sup,ejabberd_sup,<0.137.0>]
    messages: [{'$gen_event',{xmlstreamend,<<"stream:stream">>}},{'$gen_event',closed},{'DOWN',#Ref<0.0.0.63723>,process,<0.1595.0>,normal}]
    links: [<0.275.0>]
    dictionary: [{'$internal_queue_len',0},{random_seed,{5969,3781,5830}}]
    trap_exit: false
    status: running
    heap_size: 2586
    stack_size: 27
    reductions: 4120
  neighbours:
2015-12-22 12:54:18 =SUPERVISOR REPORT====
     Supervisor: {local,ejabberd_c2s_sup}
     Context:    child_terminated
     Reason:     {noproc,{gen_server,call,[<0.1595.0>,reset_stream]}}
     Offender:   [{pid,<0.1596.0>},{name,undefined},{mfargs,{ejabberd_c2s,start_link,undefined}},{restart_type,temporary},{shutdown,brutal_kill},{child_type,worker}]
```

Any ideas?

**Comment by jonathanve** [comment]
Seems it has to do with a client trying to unsuccessfully auth with mongooseim, like a dos attack. I am using now the http auth module :)

{max_fsm_queue, 1000}.

but I believe it should not kill the xmpp server

**Comment by jonathanve** [comment]
More logs

```
2015-12-30 12:28:19.069 [error] <0.18317.0> gen_fsm <0.18317.0> in state wait_for_feature_request terminated with reason: no such process or port in call to gen_server:call(<0.18316.0>, reset_stream)
2015-12-30 12:28:19.069 [error] <0.18317.0> CRASH REPORT Process <0.18317.0> with 0 neighbours exited with reason: no such process or port in call to gen_server:call(<0.18316.0>, reset_stream) in p1_fsm_old:terminate/7 line 695
2015-12-30 12:28:19.069 [error] <0.275.0> Supervisor ejabberd_c2s_sup had child undefined started with {ejabberd_c2s,start_link,undefined} at <0.18317.0> exit with reason no such process or port in call to gen_server:call(<0.18316.0>, reset_stream) in context child_terminated
```

**Comment by michalwski** [comment]
Hi @jonathanve,

Thanks for all the reports you are posting. We are going to take care of this issue in January.

**Comment by jonathanve** [comment]
thanks, @michalwski , also getting this call with blank user for http auth while the error persists:

```
GET /check_password?user=&server=x.twnel.net&pass=
```

Not sure if related.

**Comment by jonathanve** [comment]
@michalwski after including https://github.com/processone/ejabberd/commit/4819738e8f039bf08bfb9867ad279ef8a94b9b5d the crash logs disappeared; however, after 6 hours of operation mongooseim gets really slow and clients can't connect fast enough. Restarting the server is the only way to be back online.

**Comment by michalwski** [comment]
@jonathanve, many thanks for all the feedback you are giving us! 
How many users do you have connected to Mongoose and what is the memory and CPU usage after around 6h form start. Do you monitor these to see if maybe there is a huge growth? Also can you monitor metric called `[auth, backends, <<"YOUR_DOMIN_NAME">>, check_password]`. This is a histogram keeping track of check password execution time. You can find more about metrics here: http://mongooseim.readthedocs.org/en/latest/operation-and-maintenance/Mongoose-metrics/

**Comment by jonathanve** [comment]
hi @michalslaski thanks for your answer. After enabling mod_admin_extra, I discovered that the issue related with Mongooseim getting slow was a bug in our services leaving dead/zombie connections. The inclusion of https://github.com/processone/ejabberd/commit/4819738e8f039bf08bfb9867ad279ef8a94b9b5d solved the issue of mongooseim crashing.

**Comment by michalwski** [comment]
Thanks for the update. We are aware of https://github.com/processone/ejabberd/commit/4819738e8f039bf08bfb9867ad279ef8a94b9b5d, we are going to apply this change but only after we are able to reproduce similar behaviour in our test suites. Without that we cannot be sure the bug is fixed.

**Comment by michalwski** [comment]
This will be fixed in #637

**Comment by jonathanve** [comment]
thanks
